File: src/main/java/io/tripled/social/client/presentation/request/CliRequestFactory.java
package io.tripled.social.client.presentation.request;

import io.tripled.social.client.presentation.CliRequest;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class CliRequestFactory {

  private final List<RequestFactory> factories = Arrays.asList(new PostCliRequestFactory(), new ReadCliRequestFactory(),
      new FollowCliRequestFactory(), new WallCliRequestFactory());

  public Optional<CliRequest> create(String input) {
    return factories.stream()
        .filter(i -> i.supports(input))
        .findFirst()
        .map(i -> i.create(input));
  }

  interface RequestFactory {

    boolean supports(String input);

    CliRequest create(String input);

  }

  private class PostCliRequestFactory implements RequestFactory {

    @Override
    public CliRequest create(String input) {
      Matcher matcher = getMatcher(input);
      if (matcher.matches()) {
        return new PostCliRequest(matcher.group(1).trim(), matcher.group(2).trim());
      }
      return null;
    }

    @Override
    public boolean supports(String input) {
      return input != null && getMatcher(input).matches();
    }

    private Matcher getMatcher(String inputLine) {
      return Pattern.compile("(.+) -> (.+)").matcher(inputLine.trim());
    }
  }

  private class ReadCliRequestFactory implements RequestFactory {

    @Override
    public CliRequest create(String input) {
      Matcher matcher = getMatcher(input);
      if (matcher.matches()) {
        return new ReadCliRequest(matcher.group(1).trim());
      }
      return null;
    }

    @Override
    public boolean supports(String input) {
      return input != null && getMatcher(input).matches();
    }

    private Matcher getMatcher(String inputLine) {
      return Pattern.compile("(\\w+)").matcher(inputLine.trim());
    }
  }

  private class FollowCliRequestFactory implements RequestFactory {

    @Override
    public CliRequest create(String input) {
      Matcher matcher = getMatcher(input);
      if (matcher.matches()) {
        return new FollowCliRequest(matcher.group(1).trim(), matcher.group(2).trim());
      }
      return null;
    }

    @Override
    public boolean supports(String input) {
      return input != null && getMatcher(input).matches();
    }

    private Matcher getMatcher(String inputLine) {
      return Pattern.compile("(.+) follows (.+)").matcher(inputLine.trim());
    }
  }

  private class WallCliRequestFactory implements RequestFactory {

    @Override
    public CliRequest create(String input) {
      Matcher matcher = getMatcher(input);
      if (matcher.matches()) {
        return new WallCliRequest(matcher.group(1).trim());
      }
      return null;
    }

    @Override
    public boolean supports(String input) {
      return input != null && getMatcher(input).matches();
    }

    private Matcher getMatcher(String inputLine) {
      return Pattern.compile("(.+) wall").matcher(inputLine.trim());
    }
  }
}

File: src/main/kotlin/io/tripled/social/client/infrastructure/cli/request/CliRequestFactory.kt
package io.tripled.social.client.infrastructure.cli.request

import io.tripled.social.client.infrastructure.cli.CliRequest
import java.util.regex.Matcher
import java.util.regex.Pattern

class CliRequestFactory {
    private val factories: List<RequestFactory> = listOf(
        PostCliRequestFactory(), ReadCliRequestFactory(),
        FollowCliRequestFactory(), WallCliRequestFactory()
    )

    fun create(input: String): CliRequest? {
        return factories
            .filter { it.supports(input) }
            .map { it.create(input) }
            .first()

    }

    internal interface RequestFactory {
        fun supports(input: String): Boolean

        fun create(input: String): CliRequest?
    }

    private inner class PostCliRequestFactory : RequestFactory {
        override fun create(input: String): CliRequest? {
            val matcher = getMatcher(input)
            if (matcher.matches()) {
                return PostCliRequest(matcher.group(1).trim { it <= ' ' }, matcher.group(2).trim { it <= ' ' })
            }
            return null
        }

        override fun supports(input: String): Boolean {
            return getMatcher(input).matches()
        }

        fun getMatcher(inputLine: String): Matcher {
            return Pattern.compile("(.+) -> (.+)").matcher(inputLine.trim { it <= ' ' })
        }
    }

    private inner class ReadCliRequestFactory : RequestFactory {
        override fun create(input: String): CliRequest? {
            val matcher = getMatcher(input)
            if (matcher.matches()) {
                return ReadCliRequest(matcher.group(1).trim { it <= ' ' })
            }
            return null
        }

        override fun supports(input: String): Boolean {
            return getMatcher(input).matches()
        }

        fun getMatcher(inputLine: String): Matcher {
            return Pattern.compile("(\\w+)").matcher(inputLine.trim { it <= ' ' })
        }
    }

    private inner class FollowCliRequestFactory : RequestFactory {
        override fun create(input: String): CliRequest? {
            val matcher = getMatcher(input)
            if (matcher.matches()) {
                return FollowCliRequest(matcher.group(1).trim { it <= ' ' }, matcher.group(2).trim { it <= ' ' })
            }
            return null
        }

        override fun supports(input: String): Boolean {
            return getMatcher(input).matches()
        }

        fun getMatcher(inputLine: String): Matcher {
            return Pattern.compile("(.+) follows (.+)").matcher(inputLine.trim { it <= ' ' })
        }
    }

    private inner class WallCliRequestFactory : RequestFactory {
        override fun create(input: String): CliRequest? {
            val matcher = getMatcher(input)
            if (matcher.matches()) {
                return WallCliRequest(matcher.group(1).trim { it <= ' ' })
            }
            return null
        }

        override fun supports(input: String): Boolean {
            return getMatcher(input).matches()
        }

        fun getMatcher(inputLine: String): Matcher {
            return Pattern.compile("(.+) wall").matcher(inputLine.trim { it <= ' ' })
        }
    }
}
